File: cses/05_range_queries/2416_increasing-array-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    it = iter(data)
    n = int(next(it))
    q = int(next(it))

    # Đọc mảng x và tính prefix sum prefX.
    x = [0] * (n + 1)
    prefX = [0] * (n + 1)
    for i in range(1, n + 1):
        v = int(next(it))
        x[i] = v
        prefX[i] = prefX[i - 1] + v

    # Offline: nhóm các truy vấn theo điểm trái a.
    qa = [[] for _ in range(n + 1)]
    for k in range(q):
        a = int(next(it))
        b = int(next(it))
        qa[a].append((b, k))

    ans = [0] * q

    # Kỹ thuật 2 BIT cho range update + range sum: prefix(i) = s1*i - s2.
    N = n
    b1 = [0] * (n + 2)   # BIT hệ số
    b2 = [0] * (n + 2)   # BIT phần bù
    # Monotonic stack các đoạn hằng của g[i] = max(x[L..i]).
    st_val = []          # giá trị các đoạn
    st_cnt = []          # số vị trí mỗi đoạn (đỉnh = đoạn trái nhất, bắt đầu tại L+1)

    # Duyệt điểm trái L từ n về 1, thêm dần x[L] vào bên trái.
    for L in range(n, 0, -1):
        v = x[L]
        pos = L + 1
        acc = 0
        # Pop các đoạn trái nhất có value <= v và nâng chúng lên thành v.
        while st_val and st_val[-1] <= v:
            sv = st_val.pop()
            sc = st_cnt.pop()
            r = pos + sc - 1          # dải vị trí của đoạn bị pop: [pos, r]
            d = v - sv
            if d:
                # range_add(pos, r, d): _upd(pos, d), _upd(r+1, -d)
                i = pos
                vi = d * (pos - 1)
                while i <= N:
                    b1[i] += d
                    b2[i] += vi
                    i += i & (-i)
                i = r + 1
                nd = -d
                vi = nd * r
                while i <= N:
                    b1[i] += nd
                    b2[i] += vi
                    i += i & (-i)
            pos += sc
            acc += sc
        # Thêm vị trí L với giá trị v: range_add(L, L, v).
        i = L
        vi = v * (L - 1)
        while i <= N:
            b1[i] += v
            b2[i] += vi
            i += i & (-i)
        i = L + 1
        nv = -v
        vi2 = nv * L
        while i <= N:
            b1[i] += nv
            b2[i] += vi2
            i += i & (-i)
        # Đẩy đoạn mới (v, số vị trí vừa gộp + 1) vào stack.
        st_val.append(v)
        st_cnt.append(acc + 1)

        # Trả lời các truy vấn có điểm trái đúng bằng L.
        bucket = qa[L]
        if bucket:
            pxL1 = prefX[L - 1]
            Lm1 = L - 1
            for bq, k in bucket:
                # prefix(bq) = s1*bq - s2
                s1 = 0
                s2 = 0
                j = bq
                while j > 0:
                    s1 += b1[j]
                    s2 += b2[j]
                    j -= j & (-j)
                qb = s1 * bq - s2
                # prefix(L-1)
                s1 = 0
                s2 = 0
                j = Lm1
                while j > 0:
                    s1 += b1[j]
                    s2 += b2[j]
                    j -= j & (-j)
                qL = s1 * Lm1 - s2
                # Đáp án = tổng running max trên [L, bq] trừ tổng x trên [L, bq].
                ans[k] = (qb - qL) - (prefX[bq] - pxL1)

    out = "\n".join(map(str, ans))
    sys.stdout.buffer.write(out.encode())
    sys.stdout.buffer.write(b"\n")


main()

File: cses/05_range_queries/2416_increasing-array-queries/solution.cpp
#include <iostream>
#include <utility>
#include <vector>

// Fenwick / BIT hỗ trợ range update + range sum bằng kỹ thuật 2 BIT:
// một BIT cho hệ số (coefficient), một BIT cho phần bù (offset).
class RangeAddRangeSumFenwick {
public:
    explicit RangeAddRangeSumFenwick(const int size)
        : size_(size), coefficient_(size + 2, 0), offset_(size + 2, 0) {}

    // Cộng value vào mọi vị trí trong [left, right].
    void add_range(const int left, const int right, const long long value) {
        if (left > right) {
            return;
        }
        add(coefficient_, left, value);
        add(coefficient_, right + 1, -value);
        add(offset_, left, value * (left - 1));
        add(offset_, right + 1, -value * right);
    }

    // Tổng các giá trị trong [left, right].
    long long range_sum(const int left, const int right) const {
        return prefix_sum(right) - prefix_sum(left - 1);
    }

private:
    int size_;
    std::vector<long long> coefficient_;
    std::vector<long long> offset_;

    void add(std::vector<long long>& bit, int index, const long long value) {
        while (index <= size_) {
            bit[index] += value;
            index += index & -index;
        }
    }

    long long sum(const std::vector<long long>& bit, int index) const {
        long long result = 0;
        while (index > 0) {
            result += bit[index];
            index -= index & -index;
        }
        return result;
    }

    // prefix(i) = B1.query(i) * i - B2.query(i).
    long long prefix_sum(const int index) const {
        return sum(coefficient_, index) * index - sum(offset_, index);
    }
};

// Một đoạn liên tiếp mà mảng ảo g = max(x[L..i]) bằng hằng số value.
struct ConstantSegment {
    long long value;
    int length;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, q;
    std::cin >> n >> q;

    // Đọc mảng và tính prefix sum của x.
    std::vector<long long> values(n + 1, 0);
    std::vector<long long> prefix(n + 1, 0);
    for (int index = 1; index <= n; ++index) {
        std::cin >> values[index];
        prefix[index] = prefix[index - 1] + values[index];
    }

    // Offline: nhóm các truy vấn theo điểm trái (left) để xử lý một lượt.
    std::vector<std::vector<std::pair<int, int>>> queries_by_left(n + 1);
    for (int query_index = 0; query_index < q; ++query_index) {
        int left, right;
        std::cin >> left >> right;
        queries_by_left[left].emplace_back(right, query_index);
    }

    std::vector<long long> answers(q, 0);
    // Monotonic stack các đoạn hằng của g = max(x[L..i]); đỉnh là đoạn trái nhất.
    std::vector<ConstantSegment> stack;
    stack.reserve(n);
    // BIT lưu g[i] = max(x[L..i]) cho mọi i >= L (running maximum).
    RangeAddRangeSumFenwick running_maximum(n);

    // Duyệt điểm trái L từ n về 1, thêm dần x[L] vào bên trái.
    for (int left = n; left >= 1; --left) {
        const long long value = values[left];
        int next_position = left + 1;
        int merged_length = 0;

        // Pop các đoạn trái nhất có value <= x[L] và nâng chúng lên thành x[L].
        while (!stack.empty() && stack.back().value <= value) {
            const ConstantSegment segment = stack.back();
            stack.pop_back();
            const int segment_end = next_position + segment.length - 1;
            running_maximum.add_range(
                next_position, segment_end, value - segment.value
            );
            next_position = segment_end + 1;
            merged_length += segment.length;
        }

        // Thêm vị trí L với giá trị x[L] rồi đẩy đoạn mới vừa gộp vào stack.
        running_maximum.add_range(left, left, value);
        stack.push_back({value, merged_length + 1});

        // Trả lời các truy vấn có điểm trái đúng bằng L: g đang chính xác trên [L, b].
        for (const auto& [right, query_index] : queries_by_left[left]) {
            const long long maximum_sum =
                running_maximum.range_sum(left, right);
            const long long original_sum = prefix[right] - prefix[left - 1];
            answers[query_index] = maximum_sum - original_sum;
        }
    }

    for (const long long answer : answers) {
        std::cout << answer << '\n';
    }
    return 0;
}
